cli/services.go

```diff
@@ -0,0 +1,166 @@
+package cli
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-jsonrpc"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/stmgr"
+	types "github.com/filecoin-project/lotus/chain/types"
+	cid "github.com/ipfs/go-cid"
+	cbg "github.com/whyrusleeping/cbor-gen"
+	"golang.org/x/xerrors"
+)
+
+//go:generate go run github.com/golang/mock/mockgen -destination=servicesmock_test.go -package=cli -self_package github.com/filecoin-project/lotus/cli . ServicesAPI
+
+type ServicesAPI interface {
```

Comment:
Is there any way we can avoid having this additional layer? It's going to add maintenance overhead, and replicating this pattern in other places will not make adding tests easier. Does testing the command with the full API mock not work?

Reply:
> Is there any way we can avoid having this additional layer?

The point is to build up the pattern that will allow testing more complex CLI interactions.
Lotus will be getting more complex commands in near future (interactive sending, interactive deal-making as examples).
I don't think integration testing a command with interactive UI is the right way to do it.

This is the point of CLI Services, to separate the user integration and CLI parsing from actions the commands make.
This way we can clearly test the parsing or UI of the command, and we can characterise Services calls it should make.
Then we can test services themselves (if they are more complex) with mocked backend and also use the Services layer to perform very clean integration tests with real Lotus API. Using normal function calls with typed inputs and outputs, instead of serializing inputs and invoking the CLI and parsing the output.

> It's going to add maintenance overhead, and replicating this pattern in other places will not make adding tests easier.

I disagree, simple commands can be mocked out directly (`chain get head` for example) introducing this pattern does not prevent it. More complex commands will gain a lot from this pattern because of the responsibility separation allowing tests to focus on either only the parsing or behaviour.

Case in point, our hardly tested, mostly because 1. testing CLI is painful because stringifying, parsing and interacting with it; 2. no mocking

Introducing mocking is only a small step as it doesn't make writing tests for complex commands any nicer.

Reply:
It's not just about testing, it's about code structure and separation of concerns.

This separation keeps the CLI layer light and well-scoped. Its job is limited to parsing of flags/args, validating them, transforming into upper types, calling something, and presenting the information back to the user. It is no longer tasked with business logic and orchestration. The CLI layer is "dumb".

There are nice effects that arise from that:
1. Testing becomes more focused. Now you can test the behaviour of the CLI separately from the behaviour of the orchestration.
2. You can choose to expose the orchestration as a new JSON-RPC API operation.

Reply:
Also, the `send` command is right on the edge of when I would consider adding the Services layer abstraction. I wouldn't do it for much simpler commands.

Reply:
Yep, for commands that simply expose a JSON-RPC call over the CLI, you wouldn't create a services operation.

@Kubuxu how would you feel about embedding the `FullAPI` inside the full node `Services`?